### ISSUE TITLE
Ambiguity error for `DynamicalDDEProblem`

### ISSUE THREAD

**Issue by SebastianM-C** [SciML/SciMLBase.jl#42]
Running
```julia
using Test
using SciMLBase

problems = detect_ambiguities(SciMLBase)
```
gives
```julia
8-element Vector{Tuple{Method, Method}}:
 (DynamicalDDEProblem(f::DynamicalDDEFunction, h, tspan) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:77, DynamicalDDEProblem(f1::Function, f2::Function, args...; kwargs...) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:80)
 (DynamicalDDEProblem(f::DynamicalDDEFunction, h, tspan, p; kwargs...) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:77, DynamicalDDEProblem(f1::Function, f2::Function, args...; kwargs...) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:80)
 ((var"#s255"::Core.var"#Type##kw")(::Any, ::Type{DynamicalDDEProblem}, f::DynamicalDDEFunction, h, tspan, p) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:77, (::Core.var"#Type##kw")(::Any, ::Type{DynamicalDDEProblem}, f1::Function, f2::Function, args...) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:80)
 (DynamicalDDEProblem(f::DynamicalDDEFunction, v0, u0, h, tspan, p; dependent_lags, kwargs...) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:71, DynamicalDDEProblem(f1::Function, f2::Function, args...; kwargs...) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:80)
 (DynamicalDDEProblem(f::DynamicalDDEFunction, v0, u0, h, tspan) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:71, DynamicalDDEProblem(f1::Function, f2::Function, args...; kwargs...) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:80)
 ((var"#s255"::Core.var"#Type##kw")(::Any, ::Type{DynamicalDDEProblem}, f::DynamicalDDEFunction, v0, u0, h, tspan) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:71, (::Core.var"#Type##kw")(::Any, ::Type{DynamicalDDEProblem}, f1::Function, f2::Function, args...) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:80)
 ((var"#s255"::Core.var"#Type##kw")(::Any, ::Type{DynamicalDDEProblem}, f::DynamicalDDEFunction, v0, u0, h, tspan, p) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:71, (::Core.var"#Type##kw")(::Any, ::Type{DynamicalDDEProblem}, f1::Function, f2::Function, args...) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:80)
 ((var"#s255"::Core.var"#Type##kw")(::Any, ::Type{DynamicalDDEProblem}, f::DynamicalDDEFunction, h, tspan) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:77, (::Core.var"#Type##kw")(::Any, ::Type{DynamicalDDEProblem}, f1::Function, f2::Function, args...) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:80)
```

The above ambiguities refer only to two function definitions:
https://github.com/SciML/SciMLBase.jl/blob/v1.9.0/src/problems/dde_problems.jl#L77-L82

As a MWE:
```julia
using SciMLBase

f1(u,t,h) = u
f2(v,h) = v
f = DynamicalDDEFunction(f1, f2)
h(p,t) = (1,1)

DynamicalDDEProblem(f, h, (0,1))
```
which gives
```julia
ERROR: MethodError: DynamicalDDEProblem(::DynamicalDDEFunction{false, DDEFunction{false, typeof(f1), LinearAlgebra.UniformScaling{Bool}, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing}, DDEFunction{false, typeof(f2), LinearAlgebra.UniformScaling{Bool}, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing}, LinearAlgebra.UniformScaling{Bool}, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing}, ::typeof(h), ::Tuple{Int64, Int64}) is ambiguous. Candidates:
  DynamicalDDEProblem(f1::Function, f2::Function, args...; kwargs...) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:80
  DynamicalDDEProblem(f::DynamicalDDEFunction, h, tspan) in SciMLBase at C:\Users\sebastian\.julia\dev\SciMLBase\src\problems\dde_problems.jl:77
Possible fix, define
  DynamicalDDEProblem(::DynamicalDDEFunction, ::Function, ::Any)
Stacktrace:
 [1] top-level scope
```

The problem appears due to the fact that `DynamicalDDEFunction <: Function` and the commit https://github.com/SciML/SciMLBase.jl/commit/c555eada0f64fe5383f11afcf05a0d11cb4e162a removed the restriction on the `h` argument and thus in both
```julia
function DynamicalDDEProblem(f::DynamicalDDEFunction,h,tspan,p=NullParameters();kwargs...)
```
and
```julia
function DynamicalDDEProblem(f1::Function,f2::Function,args...;kwargs...)
```

the first argument fits the method, but for the second argument we have `Any` in the first case (which was more specific) and `Function` in the second (which is more specific than `Any`).

One option would be to revert the suggestion by @devmotion in https://github.com/SciML/SciMLBase.jl/commit/c555eada0f64fe5383f11afcf05a0d11cb4e162a
The other would be to replace
```julia
function DynamicalDDEProblem(f1::Function,f2::Function,args...;kwargs...)
```
with two separate methods, but this would result in code duplication as mentioned in https://github.com/SciML/SciMLBase.jl/pull/14#discussion_r567453358

I can open a PR to fix this if you would like, just let me know what approach do you think is best.

**Comment by ChrisRackauckas** [comment]
I think that could work.

**Comment by devmotion** [comment]
In general, the history function `h` does not have to be a subtype of `Function`, so adding a type constraint on `h` seems wrong.

Would it be possible to remove the type constraints on `f1` and `f2` in https://github.com/SciML/SciMLBase.jl/blob/d67b76e069fb3e9e334c137789a7cfc1ed869326/src/problems/dde_problems.jl#L80 and https://github.com/SciML/SciMLBase.jl/blob/d67b76e069fb3e9e334c137789a7cfc1ed869326/src/problems/dde_problems.jl#L100? Would that work?

**Comment by SebastianM-C** [comment]
Closed by #43
